internal/common/retry/interceptor.go

```diff
@@ -0,0 +1,146 @@
+		backoffCoefficient float64
+		maximumInterval    time.Duration
+		expirationInterval time.Duration
+		jitter             float64
+		maximumAttempts    int
+	}
+
+	contextKey string
+)
+
+// SetBackoffCoefficient sets rate at which backoff coefficient will change.
+func (g *GrpcRetryConfig) SetBackoffCoefficient(backoffCoefficient float64) {
+	g.backoffCoefficient = backoffCoefficient
+}
+
+// SetMaximumInterval defines maximum amount of time between attempts.
+func (g *GrpcRetryConfig) SetMaximumInterval(maximumInterval time.Duration) {
+	g.maximumInterval = maximumInterval
+}
+
+// SetExpirationInterval defines total amount of time that can be used for all retry attempts.
+// Note that this value is ignored if deadline is set on the context.
+func (g *GrpcRetryConfig) SetExpirationInterval(expirationInterval time.Duration) {
+	g.expirationInterval = expirationInterval
+}
+
+// SetJitter defines level of randomization for each delay interval. For example 0.2 would mex target +- 20%
+func (g *GrpcRetryConfig) SetJitter(jitter float64) {
+	g.jitter = jitter
+}
+
+// SetMaximumAttempts defines maximum total number of retry attempts.
+func (g *GrpcRetryConfig) SetMaximumAttempts(maximumAttempts int) {
+	g.maximumAttempts = maximumAttempts
+}
+
+// NewGrpcRetryConfig creates new retry config with specified initial interval and defaults for other parameters.
+// Use SetXXX functions on this config in order to customize values.
+func NewGrpcRetryConfig(initialInterval time.Duration) *GrpcRetryConfig {
+	return &GrpcRetryConfig{
+		initialInterval:    initialInterval,
+		backoffCoefficient: DefaultBackoffCoefficient,
+		maximumInterval:    DefaultMaximumInterval,
+		expirationInterval: DefaultExpirationInterval,
+		jitter:             DefaultJitter,
+		maximumAttempts:    DefaultMaximumAttempts,
+	}
+}
+
+const (
+	// ConfigKey context key for GrpcRetryConfig
+	ConfigKey = contextKey("RetryConfig")
+)
+
+// NewRetryOptionsInterceptor creates a new gRPC interceptor that populates retry options for each call based on values
+// provided in the context.
+func NewRetryOptionsInterceptor() grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		if rc, ok := ctx.Value(ConfigKey).(GrpcRetryConfig); ok {
+			if _, ok := ctx.Deadline(); !ok {
+				deadlineCtx, cancel := context.WithDeadline(ctx, time.Now().Add(rc.expirationInterval))
+				defer cancel()
+				ctx = deadlineCtx
+			}
+			opts = append(opts, grpc_retry.WithBackoff(func(attempt uint) time.Duration {
+				next := float64(rc.initialInterval) * math.Pow(rc.backoffCoefficient, float64(attempt))
+				if rc.maximumInterval != NoInterval {
+					next = math.Min(next, float64(rc.maximumInterval))
+				}
+				return backoffutils.JitterUp(time.Duration(next), rc.jitter)
+			}))
+			if rc.maximumAttempts != NoMaximumAttempts {
+				grpc_retry.WithMax(uint(rc.maximumAttempts))
```

Comment:
Maybe I'm reading this library wrong but this statement doesn't seem like it does anything? (Same with `WithCodes` below)? They return a `CallOption` but we're not using that? Seems like they need to append to `opts` like above?

Reply:
Already fixed.